### ISSUE TITLE
built-in map: busy spin loop hang of application when using a map of maps incorrectly

### ISSUE THREAD

**Issue by deckarep** [odin-lang/Odin#3265]
## Context

tldr: I failed to reference an inner map `as a reference` while trying to add keys to the inner map which causes a lockup/hang of the application to go into a busy spin loop.

Hello, it's a common pattern to have a map of maps and in my case I caused my app to go into a busy spin loop. It turns out I was doing things incorrectly and this could be regarded as user error however there was some discussion on discord that detecting this invalid behavior and crashing is better than the application hanging on a busy spin loop. tldr: I failed to pull out an item of the map `as a reference` while trying to update the inner map.

* Operating System & Odin Version:
        Odin:    dev-2024-03:8721d03cf
        OS:      macOS Ventura 22.6 (build: 22G313, kernel: 13.6.1)
        CPU:     Intel(R) Core(TM) i7-8750H CPU @ 2.20GHz
        RAM:     16384 MiB
        Backend: LLVM 14.0.6

## Expected Behavior
Perhaps a compiler error to catch such invalid cases (if possible), or the app crashing is better than it hanging silently.

## Current Behavior
The application will just hang in a busy spin-loop and use 100% of a core.

## Failure Information (for bugs)

Example code that is user error but causes the hang:

```odin
package main

import "core:fmt"

main :: proc() {
	m := make(map[string]map[string]bool)

	if "foo" not_in m {
		m["foo"] = make(map[string]bool)
	}

	elements := []string{"hello", "how", "are", "you", "doing?", "what", "time", "is", "it"}

	for s in elements {
		foo_map := m["foo"]  <-- the fix is: &m["foo"]
		foo_map[s] = true
	}


	fmt.println(m)
}

```

### Steps to Reproduce

Compile and run the code above, open activity monitor and observe a stuck core - at least this is happening on Darwin.

### Failure Logs

This is a sample profile of around where the busy-spin loop is choking.

```
Call graph:
    2440 Thread_26449070   DispatchQueue_1: com.apple.main-thread  (serial)
      1977 runtime.map_insert_hash_dynamic  (in simple_map_bug) + 3293,3576,...  [0x10e0100fd,0x10e010218,...]
      121 runtime.map_insert_hash_dynamic  (in simple_map_bug) + 2078  [0x10e00fc3e]
      + 121 runtime.map_hash_is_deleted  (in simple_map_bug) + 25,12,...  [0x10e00e219,0x10e00e20c,...]
      70 runtime.map_insert_hash_dynamic  (in simple_map_bug) + 3757  [0x10e0102cd]
      + 52 _platform_memmove$VARIANT$Haswell  (in libsystem_platform.dylib) + 71,185,...  [0x7ff8111c8a07,0x7ff8111c8a79,...]
      + 9 DYLD-STUB$$_platform_memmove  (in libsystem_platform.dylib) + 0  [0x7ff8111cf57a]
      + 9 DYLD-STUB$$memcpy  (in simple_map_bug) + 0  [0x10e044a74]
      62 runtime.map_insert_hash_dynamic  (in simple_map_bug) + 3815  [0x10e010307]
      + 57 _platform_memmove$VARIANT$Haswell  (in libsystem_platform.dylib) + 185,7,...  [0x7ff8111c8a79,0x7ff8111c89c7,...]
      + 5 DYLD-STUB$$_platform_memmove  (in libsystem_platform.dylib) + 0  [0x7ff8111cf57a]
      61 runtime.map_insert_hash_dynamic  (in simple_map_bug) + 3728  [0x10e0102b0]
      + 46 _platform_memmove$VARIANT$Haswell  (in libsystem_platform.dylib) + 163,179,...  [0x7ff8111c8a63,0x7ff8111c8a73,...]
      + 9 DYLD-STUB$$_platform_memmove  (in libsystem_platform.dylib) + 0  [0x7ff8111cf57a]
      + 6 DYLD-STUB$$memcpy  (in simple_map_bug) + 0  [0x10e044a74]
      56 runtime.map_insert_hash_dynamic  (in simple_map_bug) + 3699  [0x10e010293]
      + 40 _platform_memmove$VARIANT$Haswell  (in libsystem_platform.dylib) + 71,7,...  [0x7ff8111c8a07,0x7ff8111c89c7,...]
      + 11 DYLD-STUB$$_platform_memmove  (in libsystem_platform.dylib) + 0  [0x7ff8111cf57a]
      + 5 DYLD-STUB$$memcpy  (in simple_map_bug) + 0  [0x10e044a74]
      49 runtime.map_insert_hash_dynamic  (in simple_map_bug) + 3670  [0x10e010276]
      + 38 _platform_memmove$VARIANT$Haswell  (in libsystem_platform.dylib) + 163,7,...  [0x7ff8111c8a63,0x7ff8111c89c7,...]
      + 11 DYLD-STUB$$_platform_memmove  (in libsystem_platform.dylib) + 0  [0x7ff8111cf57a]
      44 runtime.map_insert_hash_dynamic  (in simple_map_bug) + 3786  [0x10e0102ea]
        34 _platform_memmove$VARIANT$Haswell  (in libsystem_platform.dylib) + 7,202,...  [0x7ff8111c89c7,0x7ff8111c8a8a,...]
        10 DYLD-STUB$$_platform_memmove  (in libsystem_platform.dylib) + 0  [0x7ff8111cf57a]

Total number in stack (recursive counted multiple, when >=5):
        6       DYLD-STUB$$_platform_memmove  (in libsystem_platform.dylib) + 0  [0x7ff8111cf57a]
        6       _platform_memmove$VARIANT$Haswell  (in libsystem_platform.dylib) + 0  [0x7ff8111c89c0]

Sort by top of stack, same collapsed (when >= 5):
        runtime.map_insert_hash_dynamic  (in simple_map_bug)        1977
        _platform_memmove$VARIANT$Haswell  (in libsystem_platform.dylib)        267
        runtime.map_hash_is_deleted  (in simple_map_bug)        121
        DYLD-STUB$$_platform_memmove  (in libsystem_platform.dylib)        55
        DYLD-STUB$$memcpy  (in simple_map_bug)        20
```

**Comment by gingerBill** [comment]
> Perhaps a compiler error to catch such invalid cases (if possible)

It's probably not possible to solve this in general because the compiler cannot know what you were trying to do.

**Comment by JesseRMeyer** [comment]
But surely an infinite loop in the hash lookup is undesirable?

**Comment by gingerBill** [comment]
The problem is that the original data is using the wrong memory whilst the copy has been updated.

**Comment by gingerBill** [comment]
This should have been fixed now with a `panic`
